invenio/modules/deposit/models.py

```diff
@@ -521,8 +553,29 @@ def __unicode__(self):
         """ Return a name for this class """
         return self.get_identifier()
 
+    @classmethod
+    def all_authorized(cls, current_user):
+        """
+        Return a dict of deposit types that the current user
+        is allowed to view.
+        """
+        from invenio.modules.access.control import acc_get_action_id, \
+            acc_find_possible_actions_user
+
+        all_types = cls.all()
+        action_id = acc_get_action_id('usedeposit')
+
+        avail_actions = acc_find_possible_actions_user(current_user.get_id(),
+                                                       action_id)[1:]
+        values = [s for s in itertools.chain(*avail_actions)
+                  if isinstance(s, basestring)]
+
+        # filter the all_types dict()
+        return {key: all_types[key] for key in values if key in all_types}
```

Comment:
> Can you also explain me how do you calculate the `precached_allowed_deposition_types`?

Please check: `invenio.ext.login.legacy_user`

> I think this is semantically the same as code above

You can check if one of the authorized action has `type=None` then **all** depositions should be allowed.

```python
avail_actions = acc_find_possible_actions_user(
    current_user.get_id(),
    action_id)[1:]
values = set(s for s in itertools.chain(*avail_actions) if isinstance(s, basestring))
# check 
if None in values:
    # return all actions
return # filter allowed actions
```